graph-builder/src/config/settings.rs

```diff
@@ -0,0 +1,92 @@
+use super::Merge;
+use super::{cli, file};
+use failure::Fallible;
+use std::collections::HashSet;
+use std::net::{IpAddr, Ipv4Addr};
+use std::path::PathBuf;
+use std::time;
+use structopt::StructOpt;
+
+/// Runtime application settings (validated config).
+#[derive(Debug, SmartDefault)]
+pub struct AppSettings {
+    /// Listening address for the main service.
+    #[default(IpAddr::V4(Ipv4Addr::LOCALHOST))]
+    pub address: IpAddr,
+
+    /// Optional auth secrets for the registry scraper.
+    #[default(None)]
+    pub credentials_path: Option<PathBuf>,
+
+    /// Required client parameters for the main service.
+    #[default(_code = "HashSet::new()")]
```

Comment:
The comment below will probably influence this, but in the current stat we could use `Default::default()` here